libraries/pod/src/option.rs

```diff
@@ -0,0 +1,140 @@
+//! Generic `Option` that can be used as a `Pod` for types that can have
+//! a designated `None` value.
+//!
+//! For example, a 64-bit unsigned integer can designate `0` as a `None` value.
+//! This would be equivalent to
+//! [`Option<NonZeroU64>`](https://doc.rust-lang.org/std/num/type.NonZeroU64.html)
+//! and provide the same memory layout optimization.
+
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_program::{program_option::COption, pubkey::Pubkey},
+};
+
+/// Trait for types that can be `None`.
+///
+/// This trait is used to indicate that a type can be `None` according to a
+/// specific value.
+pub trait Nullable: Pod {
+    /// Indicates whether the value is `None` or not.
+    fn is_none(&self) -> bool;
+}
+
+/// A "pod-enabled" type that can be used as an `Option<T>` without
+/// requiring extra space to indicate if the value is `Some` or `None`.
+///
+/// This can be used when a specific value of `T` indicates that its
+/// value is `None`.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct PodOption<T: Nullable>(T);
+
+impl<T: Nullable> PodOption<T> {
+    /// Returns the contained value as an `Option`.
+    pub fn get(self) -> Option<T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(self.0)
+        }
+    }
+}
+
+unsafe impl<T: Nullable> Pod for PodOption<T> {}
+
+unsafe impl<T: Nullable> Zeroable for PodOption<T> {}
+
+impl<T: Nullable> From<T> for PodOption<T> {
+    fn from(value: T) -> Self {
+        PodOption(value)
+    }
+}
+
+impl<T: Nullable> From<PodOption<T>> for Option<T> {
+    fn from(from: PodOption<T>) -> Self {
+        from.get()
+    }
+}
+
+impl<'a, T: Nullable> From<&'a PodOption<T>> for Option<&'a T> {
+    fn from(from: &'a PodOption<T>) -> Self {
+        if from.0.is_none() {
+            None
+        } else {
+            Some(&from.0)
+        }
+    }
+}
+
+impl<'a, T: Nullable> From<&'a mut PodOption<T>> for Option<&'a mut T> {
+    fn from(from: &'a mut PodOption<T>) -> Self {
+        if from.0.is_none() {
+            None
+        } else {
+            Some(&mut from.0)
+        }
+    }
+}
+
+impl<T: Nullable> From<PodOption<T>> for COption<T> {
+    fn from(from: PodOption<T>) -> Self {
+        if from.0.is_none() {
+            COption::None
+        } else {
+            COption::Some(from.0)
+        }
+    }
+}
+
+impl<'a, T: Nullable> From<&'a PodOption<T>> for COption<&'a T> {
+    fn from(from: &'a PodOption<T>) -> Self {
+        if from.0.is_none() {
+            COption::None
+        } else {
+            COption::Some(&from.0)
+        }
+    }
+}
```

Comment:
No need for this, since the user can just call `.into().as_ref()`!

https://github.com/anza-xyz/agave/blob/8db8e60c48ab064c88a76013597f99c9eb25ed74/sdk/program/src/program_option.rs#L126-L131